### DISCUSSION TITLE
dynamic import without tree shaking

### DISCUSSION THREAD

**Discussion by heart-er** [webpack/webpack#15052] [Q&A]
OS: MacOS
Webpack: 5.65.0

What is the current behavior?
dynamic import module not tree shaking about unused variable

**example**
main.js
![image](https://user-images.githubusercontent.com/24717511/147356752-ee2f8a9a-4fb9-424c-abde-ed9d71359e00.png)

a.js
![image](https://user-images.githubusercontent.com/24717511/147356773-afd2e4af-8150-422d-8a5b-6681671aa647.png)

b.js
![image](https://user-images.githubusercontent.com/24717511/147356739-92484b89-a49f-4b2e-a29b-5a8d395ce65a.png)

output
![image](https://user-images.githubusercontent.com/24717511/147356797-cd20326a-51c2-4623-aaa3-ff6b8010c96d.png)

test is unused variable, it's not tree shaking.

**Comment by alexander-akait** [comment]
Because dynamic import can have side effects, anyway terser remove should remove this code, minimize your code

**Reply by heart-er** [reply]
but in my demo, i have add terser plugin, but terser not remove unused code.

**Reply by alexander-akait** [reply]
You provide not minimizer code

**Reply by alexander-akait** [reply]
If you provide reproducible test repo I will show how to achieve it

**Reply by heart-er** [reply]
ok, repo comming!
https://github.com/heart-er/webpack-test-repo
The repo is private, i have invited you to join repo.

**Reply by alexander-akait** [reply]
Thanks, I will look at this in near future (today/tomorrow)

**Reply by heart-er** [reply]
Oh, i think i has achieve it, because i commented terser before, so it's not tree shaking.
![image](https://user-images.githubusercontent.com/24717511/147358387-0dc9fe46-7621-4bfc-918b-b4a1125fbbf2.png)

My output code has tree shaking by adding Terser plugin

Thank you very much!

**Reply by alexander-akait** [reply]
Yep, webpack doesn't remove code, just leaving it in code as unused and terser removed unused code

**Reply by nyngwang** [reply]
@alexander-akait: how to append more terserOptions?

**Reply by vankop** [reply]
https://webpack.js.org/plugins/terser-webpack-plugin/
